In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
df = pd.read_csv(r"C:\Users\Prerak\Desktop\Credit Card Fraud Detection\creditcard.csv.crdownload")

In [7]:
df

Time        V1        V2        V3        V4        V5        V6  \
0          0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1          0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2          1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3          1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4          2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...      ...       ...       ...       ...       ...       ...       ...   
11660  19915  1.294875 -0.645847  0.689549 -0.351634 -1.026884 -0.129811   
11661  19915  1.404683 -0.554883  0.612239 -0.234956 -1.190992 -0.816824   
11662  19915 -0.945541  0.479754  1.521916 -1.298658 -0.852548 -0.604029   
11663  19915 -0.087909  0.184093  1.683910 -0.837378 -0.682605 -0.669907   
11664  19915  1.504229 -0.499337  0.052377 -0.576345 -0.393971  0.015149   

             V7        V8        V9  ...       V21       V22       V23  \
0      0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1     -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2      0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3      0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4      0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...         ...       ...       ...  ...       ...       ...       ...   
11660 -0.928101  0.114172  0.804717  ... -0.097462 -0.020893  0.040297   
11661 -0.775771 -0.142637  1.121638  ... -0.196110 -0.303562  0.005661   
11662 -0.354686  0.498106  0.407159  ... -0.050745  0.056031 -0.055584   
11663 -0.056222 -0.120669  0.409636  ... -0.109428 -0.054760 -0.139329   
11664 -0.718603 -0.125852  1.023281  ...       NaN       NaN       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
0      0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62    0.0  
1     -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69    0.0  
2     -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66    0.0  
3     -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50    0.0  
4      0.141267 -0.206010  0.502292  0.219422  0.215153   69.99    0.0  
...         ...       ...       ...       ...       ...     ...    ...  
11660  0.159282  0.350705 -0.278351 -0.010354 -0.008003   10.00    0.0  
11661  0.313352  0.473813 -0.276618 -0.024026  0.002106    5.00    0.0  
11662  0.312195  0.026299 -0.413466  0.189152  0.086360   10.00    0.0  
11663  0.333267  0.072695 -0.320292  0.006423  0.010148   10.00    0.0  
11664       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[11665 rows x 31 columns]

In [8]:
#First we check null in out data or not
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [9]:
df = df.dropna()
# drop the null data 

In [10]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [11]:
df['Amount'].describe()

count    11664.000000
mean        62.795132
std        179.816783
min          0.000000
25%          5.000000
50%         15.950000
75%         50.000000
max       7712.430000
Name: Amount, dtype: float64

In [12]:
Fraud =print(f'Number of Fraud transactions: {df[df["Class"] == 1.0].shape[0]}')
Genuine = print(f'Number of Genuine transactions: {df[df["Class"] == 0.0].shape[0]}')




Number of Fraud transactions: 49
Number of Genuine transactions: 11615


In [13]:
(df[df["Class"] == 1.0].shape[0])/(df[df["Class"] == 0.0].shape[0])*100

0.4218682737839001

0.42 percentage of fraud transaction

In [14]:
from sklearn.preprocessing import StandardScaler


Normalize the data by using a stand-alone scaler. Why? Normalization is used to assign all values to a scale between 1 and 0, which prevents bias, speeds processing, and reduces overfitting.

In [15]:
s = StandardScaler()

In [16]:
df['Amount'] = s.fit_transform(df['Amount'].values.reshape(-1, 1))
# transforming the data

C:\Users\Prerak\AppData\Local\Temp\ipykernel_24944\2290020765.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Amount'] = s.fit_transform(df['Amount'].values.reshape(-1, 1))


In [17]:
df = df.drop(columns=['Time'])
#drop the time column beacuse dont use for prediction

In [18]:
df

V1        V2        V3        V4        V5        V6        V7  \
0     -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1      1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
...         ...       ...       ...       ...       ...       ...       ...   
11659  1.348167 -0.531813  0.737667 -0.312410 -1.147767 -0.585711 -0.810412   
11660  1.294875 -0.645847  0.689549 -0.351634 -1.026884 -0.129811 -0.928101   
11661  1.404683 -0.554883  0.612239 -0.234956 -1.190992 -0.816824 -0.775771   
11662 -0.945541  0.479754  1.521916 -1.298658 -0.852548 -0.604029 -0.354686   
11663 -0.087909  0.184093  1.683910 -0.837378 -0.682605 -0.669907 -0.056222   

             V8        V9       V10  ...       V21       V22       V23  \
0      0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474   
1      0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288   
2      0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412   
3      0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321   
4     -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458   
...         ...       ...       ...  ...       ...       ...       ...   
11659 -0.062116  0.705208  0.317810  ... -0.120215 -0.076450  0.006995   
11660  0.114172  0.804717  0.291495  ... -0.097462 -0.020893  0.040297   
11661 -0.142637  1.121638  0.176953  ... -0.196110 -0.303562  0.005661   
11662  0.498106  0.407159 -0.608412  ... -0.050745  0.056031 -0.055584   
11663 -0.120669  0.409636 -0.285711  ... -0.109428 -0.054760 -0.139329   

            V24       V25       V26       V27       V28    Amount  Class  
0      0.066928  0.128539 -0.189115  0.133558 -0.021053  0.482873    0.0  
1     -0.339846  0.167170  0.125895 -0.008983  0.014724 -0.334272    0.0  
2     -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  1.756668    0.0  
3     -1.175575  0.647376 -0.221929  0.062723  0.061458  0.337607    0.0  
4      0.141267 -0.206010  0.502292  0.219422  0.215153  0.040014    0.0  
...         ...       ...       ...       ...       ...       ...    ...  
11659  0.489243  0.448523 -0.309831 -0.021411 -0.003987 -0.321425    0.0  
11660  0.159282  0.350705 -0.278351 -0.010354 -0.008003 -0.293618    0.0  
11661  0.313352  0.473813 -0.276618 -0.024026  0.002106 -0.321425    0.0  
11662  0.312195  0.026299 -0.413466  0.189152  0.086360 -0.293618    0.0  
11663  0.333267  0.072695 -0.320292  0.006423  0.010148 -0.293618    0.0  

[11664 rows x 30 columns]

In [19]:
from sklearn.model_selection import train_test_split
#split the data 

In [20]:
x = df.iloc[:,:-1]
y = df[['Class']]
# x is independent column 
# y is dependent column

In [21]:
x

V1        V2        V3        V4        V5        V6        V7  \
0     -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1      1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
...         ...       ...       ...       ...       ...       ...       ...   
11659  1.348167 -0.531813  0.737667 -0.312410 -1.147767 -0.585711 -0.810412   
11660  1.294875 -0.645847  0.689549 -0.351634 -1.026884 -0.129811 -0.928101   
11661  1.404683 -0.554883  0.612239 -0.234956 -1.190992 -0.816824 -0.775771   
11662 -0.945541  0.479754  1.521916 -1.298658 -0.852548 -0.604029 -0.354686   
11663 -0.087909  0.184093  1.683910 -0.837378 -0.682605 -0.669907 -0.056222   

             V8        V9       V10  ...       V20       V21       V22  \
0      0.098698  0.363787  0.090794  ...  0.251412 -0.018307  0.277838   
1      0.085102 -0.255425 -0.166974  ... -0.069083 -0.225775 -0.638672   
2      0.247676 -1.514654  0.207643  ...  0.524980  0.247998  0.771679   
3      0.377436 -1.387024 -0.054952  ... -0.208038 -0.108300  0.005274   
4     -0.270533  0.817739  0.753074  ...  0.408542 -0.009431  0.798278   
...         ...       ...       ...  ...       ...       ...       ...   
11659 -0.062116  0.705208  0.317810  ... -0.060712 -0.120215 -0.076450   
11660  0.114172  0.804717  0.291495  ... -0.126197 -0.097462 -0.020893   
11661 -0.142637  1.121638  0.176953  ... -0.125651 -0.196110 -0.303562   
11662  0.498106  0.407159 -0.608412  ... -0.106545 -0.050745  0.056031   
11663 -0.120669  0.409636 -0.285711  ...  0.003762 -0.109428 -0.054760   

            V23       V24       V25       V26       V27       V28    Amount  
0     -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  0.482873  
1      0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724 -0.334272  
2      0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  1.756668  
3     -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  0.337607  
4     -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153  0.040014  
...         ...       ...       ...       ...       ...       ...       ...  
11659  0.006995  0.489243  0.448523 -0.309831 -0.021411 -0.003987 -0.321425  
11660  0.040297  0.159282  0.350705 -0.278351 -0.010354 -0.008003 -0.293618  
11661  0.005661  0.313352  0.473813 -0.276618 -0.024026  0.002106 -0.321425  
11662 -0.055584  0.312195  0.026299 -0.413466  0.189152  0.086360 -0.293618  
11663 -0.139329  0.333267  0.072695 -0.320292  0.006423  0.010148 -0.293618  

[11664 rows x 29 columns]

In [22]:
y

Class
0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
...      ...
11659    0.0
11660    0.0
11661    0.0
11662    0.0
11663    0.0

[11664 rows x 1 columns]

In [23]:

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=22)

In [24]:
from sklearn.linear_model import LogisticRegression

In [25]:
lo = LogisticRegression()

In [26]:
model = lo.fit(x_train,y_train)

C:\Users\Prerak\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1184: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Prerak\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [27]:
prediction = model.predict(x_test)

In [42]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [29]:
accuracy_score(prediction,y_test)

0.9991428571428571

In [43]:
confusion_matrix(prediction,y_test)

array([[3484,    2],
       [   1,   13]], dtype=int64)

In [30]:
# logical regression get 99% accurancy 


### Dicision Tress

let's try decision tress

In [31]:
from sklearn.tree import DecisionTreeClassifier

In [32]:
de = DecisionTreeClassifier()

In [33]:
model2 = de.fit(x_train,y_train)

In [34]:
prediction2 = model2.predict(x_test)

In [35]:
accuracy_score(prediction2,y_test)

0.9994285714285714

In [44]:
confusion_matrix(prediction2,y_test)

array([[3483,    0],
       [   2,   15]], dtype=int64)

True Negative (TN): 3483 (instances correctly predicted as negative)

False Positive (FP): 0 (instances incorrectly predicted as positive)

False Negative (FN): 2 (instances incorrectly predicted as negative)

True Positive (TP): 15 (instances correctly predicted as positive)

 We get almost the same accuracy.

### Random Forest 

In [37]:
from sklearn.ensemble import RandomForestClassifier

In [38]:
r = RandomForestClassifier(n_estimators=101)

In [39]:
model3 = r.fit(x_train,y_train)

C:\Users\Prerak\anaconda3\Lib\site-packages\sklearn\base.py:1151: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


In [40]:
prediction3 = model3.predict(x_test)

In [41]:
accuracy_score(prediction3,y_test)

0.9997142857142857

In [45]:
confusion_matrix(prediction3,y_test)

array([[3484,    0],
       [   1,   15]], dtype=int64)

True Negative (TN): 3483 (instances correctly predicted as negative)

False Positive (FP): 0 (instances incorrectly predicted as positive)

False Negative (FN): 1 (instances incorrectly predicted as negative)

True Positive (TP): 15 (instances correctly predicted as positive)

######    Our accuracy is almost identical.